src/w25n.rs

```diff
@@ -0,0 +1,229 @@
+use crate::{BlockDevice, Error, Read};
+use crate::w25m::Stackable;
+use embedded_hal::blocking::spi::Transfer;
+use embedded_hal::digital::v2::OutputPin;
+use bitflags::bitflags;
+use core::fmt::Debug;
+use core::convert::TryInto;
+
+enum Opcode {
+    // Read one of the 3 8 bit status registers
+    ReadStatus = 0x05,
+    // Set the write enable latch.
+    WriteEnable = 0x06,
+    // Write to one of the three status registers
+    WriteStatus = 0x1F,
+    // Erase a 128 kb block
+    BlockErase = 0xD8,
+    // Read one page of data into the buffer
+    PageDataRead = 0x13,
+    // Read data from the buffer
+    ReadData = 0x03,
+    // Write a page of data from the buffer into a memory region
+    ProgramExecute = 0x10,
+    // Write a page of data into the buffer
+    LoadProgramData = 0x02,
+}
+
+bitflags! {
+    /// Status register bits.
+    pub struct Status3: u8 {
+        /// Erase or write in progress.
+        const BUSY = 1 << 0;
+        /// Status of the **W**rite **E**nable **L**atch.
+        const WEL = 1 << 1;
+    }
+}
+
+/// Driver for W25N series SPI Flash chips.
+///
+/// # Type Parameters
+///
+/// * **`SPI`**: The SPI master to which the flash chip is attached.
+/// * **`CS`**: The **C**hip-**S**elect line attached to the `\CS`/`\CE` pin of
+///   the flash chip.
+#[derive(Debug)]
+pub struct Flash<SPI: Transfer<u8>, CS: OutputPin> {
+    spi: SPI,
+    cs: CS,
+}
+
+impl<SPI: Transfer<u8>, CS: OutputPin> Flash<SPI, CS> {
+    /// Creates a new 25-series flash driver.
+    ///
+    /// # Parameters
+    ///
+    /// * **`spi`**: An SPI master. Must be configured to operate in the correct
+    ///   mode for the device.
+    /// * **`cs`**: The **C**hip-**S**elect Pin connected to the `\CS`/`\CE` pin
+    ///   of the flash chip. Will be driven low when accessing the device.
+    pub fn init(spi: SPI, cs: CS) -> Result<Self, Error<SPI, CS>> {
+        let mut this = Self { spi, cs };
+        let status = this.read_status_3()?;
+        info!("Flash::init: status = {:?}", status);
+        // Here we don't expect any writes to be in progress, and the latch must
+        // also be deasserted.
+        if !(status & (Status3::BUSY | Status3::WEL)).is_empty() {
+            return Err(Error::UnexpectedStatus);
+        }
+
+        // write to config register 2, set BUF=0 (continious mode) and everything else on reset
+        this.command(&mut [Opcode::WriteStatus as u8, 0xA0, 0b00000010])?;
+        this.command(&mut [Opcode::WriteStatus as u8, 0xB0, 0b00010000])?;
```

Comment:
Hmm, could this perhaps be integrated into the w25m `Flash`? Then we wouldn't have to duplicate the series25 module.

Reply:
Yeah we still have to duplicate it because of the method the w25n chips use to read and write based on their internal buffer which is a thing the series25 chips are not doing at the moment